### PR TITLE
Bug 2021629: UPSTREAM: <carry>: api request counts for current hour are incorrect

### PR DESCRIPTION
**Pull request by sanchezl**
Customers upgrading from 4.7 to 4.8 noticed that `APIRequestCount.status.currentHour` for contained counts for deprecated APIs, whose use by OpenShift was removed in 4.8, long after the upgrade to 4.8 had completed.

When the apiserver is restarted, it looses track of the `APIRequestCount` resources it was working on before the restart, and as a result, they are never updated again if the corresponding API is never used again. This situation is exacerbated during an upgrade as components are upgraded and no longer use the older APIs and the apiservers themselves are restarted. 

- Discover existing resource API request count logs on apiserver restart.
- Prevent un-needed updates.
- Fixup small typos
- Reduce chatter in apiserver logs

### CODE REVIEW

**Comment by deads2k** [comment]
Does this clear every hour for that resource or just the current one.  Are our unit tests good enough to describe this case?

**Comment by sanchezl** [comment]
> Does this clear every hour for that resource or just the current one. Are our unit tests good enough to describe this case?

This will always clear (actually, updated to `status.last24h[currentHour]`) the `status.currentHour` field.

Expired hours in `status.last24h[]` that were not cleared because the node was down for over an hour, will eventually clear 24 hours later.

More unit tests added.

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go:641]: if we ran this test for each hour leading back to 20, the idea is that I should see this cleared, right?

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go:639]: This is what the PR does, right?

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go:120]: explain why this works for the next hour too.

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go:122]: There is a side-effect of this that means that it becomes impossible to remove apicounts.  Practically, that's a very unlikely problem, but there needs to be a plan (manual with etcdctl I suspect) to get someone unstuck.

1. 4.9 starts at t=0
2. access ingress.v1beta1
3. write apicount at t0+30m
4. 4.9 stops
5. 4.10 starts t0+2h
6. nobody ever touchest ingress.v1beta1 again
7. t0+6h, cluster-admin deletes apicount/ingress.v1beta1
8. apicount/ingress.v1beta1 gets recreated (before this PR, it did not get recreated)

**Comment by deads2k** [comment]
lgtm after the comments are addressed.

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go:120]: Updated doc.

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller.go:125]: I have a removed the side-effect (step 8 should not happen). See `ApplyStatus()` in `deprecatedapirequest/v1helpers/helpers.go`.

File: openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/apiaccess_count_controller_test.go:640]: Added unit test.

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go:36]: doesn't this `apiRequestCount` that you're updating here for new values end up overwritten below?  Is the re-use ok because we stored the output in `newStatus`?

Can we just rename the one below so that readers like me dont' get confused?

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go:49]: The reasoning is good, but I don't see the check for "does the updated status have counts in it"

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (deads2k) [inline openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go:44]: Looks like past me re-used the same variable too this too.  He did a bad thing.

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go:49]: New requests would resulted in an updated status and gotten us past the `equality.Semantic.DeepEqual(&oldStatus, newStatus)` call above above.

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go]: Done.

File: openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go
Comment (sanchezl) [reply openshift-kube-apiserver/filters/deprecatedapirequest/v1helpers/helpers.go:44]: Updated.

**Comment by deads2k** [comment]
/hold

for @sttts

**Review by soltysh — APPROVED** [review]
/lgtm

**Comment by soltysh** [comment]
/hold
go1.17 caused problems, I'll fix that in a separate PR

**Pull request merged**
